backend/src/nodes/nodes/external_stable_diffusion/img2img.py

```diff
@@ -0,0 +1,104 @@
+    EnumInput,
+    ImageInput,
+)
+from ...properties.outputs import ImageOutput
+from typing import Optional
+
+
+@NodeFactory.register("chainner:external_stable_diffusion:img2img")
+class Img2Img(NodeBase):
+    def __init__(self):
+        super().__init__()
+        self.description = "Modify an image using Automatic1111"
+        self.inputs = [
+            ImageInput(),
+            TextInput("Prompt", default="an astronaut riding a horse"),
+            TextInput("Negative Prompt").make_optional(),
+            SliderInput(
+                "Denoising Strength",
+                minimum=0,
+                default=0.75,
+                maximum=1,
+                slider_step=0.01,
+                controls_step=0.1,
+                precision=2,
+            ),
+            group("seed")(
+                NumberInput("Seed", minimum=0, default=42, maximum=4294967296)
+            ),
+            SliderInput("Steps", minimum=1, default=20, maximum=150),
+            EnumInput(SamplerName, default_value=SamplerName.EULER),
+            SliderInput(
+                "CFG Scale",
+                minimum=1,
+                default=7,
+                maximum=20,
+                controls_step=0.1,
+                precision=1,
+            ),
+            SliderInput("Width", minimum=64, default=512, maximum=2048).with_id(8),
+            SliderInput("Height", minimum=64, default=512, maximum=2048).with_id(9),
+            TextInput("Model Checkpoint Override").make_optional(),
+        ]
+        self.outputs = [
+            ImageOutput(
+                image_type="Image {width: Input8, height: Input9, channels: 3}"
+            ),
+        ]
+
+        self.category = ExternalStableDiffusionCategory
+        self.name = "Image to Image"
+        self.icon = "MdChangeCircle"
+        self.sub = "Automatic1111"
+
+    def run(
+        self,
+        image: np.ndarray,
+        prompt: str,
+        negative_prompt: Optional[str],
+        denoising_strength: float,
+        seed: int,
+        steps: int,
+        sampler_name: SamplerName,
+        cfg_scale: float,
+        width: int,
+        height: int,
+        sd_model_checkpoint: Optional[str],
+    ) -> np.ndarray:
+        request_data = {
+            "init_images": [encode_base64_image(image)],
+            "prompt": prompt,
+            "negative_prompt": negative_prompt or "",
+            "denoising_strength": denoising_strength,
+            "seed": seed,
+            "steps": steps,
+            "sampler_name": sampler_name.value,
+            "cfg_scale": cfg_scale,
+            "width": width,
+            "height": height,
+            "override_settings": {},
+        }
+        if sd_model_checkpoint:
+            request_data["override_settings"][
+                "sd_model_checkpoint"
+            ] = sd_model_checkpoint
+        response = post(url=STABLE_DIFFUSION_IMG2IMG_URL, json_data=request_data)
+        return decode_base64_image(response["images"][0])
```

Comment:
As with channels, we should also have a runtime check here to verify the size of the returned image.

Same for text to image.
```suggestion
        result = decode_base64_image(response["images"][0])
        h, w, _ = get_h_w_c(result)
        assert (w, h) == (width, height), f"Expected the returned image to be {width}x{height}px but found {w}x{h}px instead"
        return result
```

Reply:
Cool!  Yes, makes sense.

> Sorry about this.

No worries!  I appreciate you taking the time to review it.

Reply:
Actually, this has raised a weird issue that I missed.  The API does not in fact always return an image of the size you request.  I thought I'd tested this, but I guess I happened to only try multiple of 8, because actually it seems to round down.

If you ask for 513x513, it will return 512x512.  If you ask for 511x511 it will return 504x504.

I added `control_step` and `slider_step` arguments to the SliderInputs, but it doesn't stop someone from manually setting an invalid size.

https://github.com/chaiNNer-org/chaiNNer/blob/059bb95a1fef73b30935a25dd8624dcc360fa6cf/backend/src/nodes/nodes/external_stable_diffusion/img2img.py#L60-L67

For now I added this runtime error, but I think it'd be better if we could predict the output size instead.  Is it possible, I wonder, to specify this "round down" behavior in the image type of the ImageOutput?  I get lost when I try to find out where things are happening in the typescript code.

https://github.com/chaiNNer-org/chaiNNer/blob/059bb95a1fef73b30935a25dd8624dcc360fa6cf/backend/src/nodes/nodes/external_stable_diffusion/img2img.py#L101-L104

Reply:
> For now I added this runtime error, but I think it'd be better if we could predict the output size instead. Is it possible, I wonder, to specify this "round down" behavior in the image type of the ImageOutput? I get lost when I try to find out where things are happening in the typescript code.

Navi docs are [here](https://github.com/chaiNNer-org/Navi/blob/main/navi.md#built-in-functions), but it seems like I forgot to document the `floor` and `ceil` functions. I'll fix that soon.
And yes, there is a lot of TS code for the type system. It's a whole little programming language :)

Anyway, yes, we can adjust the output type to do rounding to get rid of the runtime error. Since we have a `floor` function, we can do `int & floor(InputX / 8) * 8` to round down to the nearest multiple of 8.

If you want to, you can even extract this into a function:
```
def nearest_valid(n: number) = int & floor(n / 8) * 8;
Image { 
    width: nearest_valid(InputX),
    height: nearest_valid(InputY)
}
```

Reply:
Oh, awesome!  I didn't realize there was documentation!  Thank you.  I've put that in and it seems to be working.